Sources/TuistCache/ContentHashing/GraphContentHasher.swift

```diff
@@ -2,78 +2,88 @@ import Checksum
-        let platformHash = try hash(string: targetNode.target.platform.rawValue)
-        return try hash(strings: [sourcesHash, productHash, platformHash])
-    }
-
-    fileprivate func hashSources(of targetNode: TargetNode) throws -> String {
-        let hashes = try targetNode.target.sources.sorted(by: { $0.path < $1.path }).map(md5)
-        let joinedHash = try hash(strings: hashes)
-        return joinedHash
-    }
+    private func hash(targetNode: TargetNode) throws -> String {
+        let target = targetNode.target
+        let sourcesHash = try sourceFilesContentHasher.hash(sources: target.sources)
+        let resourcesHash = try hash(resources: target.resources)
+        let coreDataModelHash = try coreDataModelsContentHasher.hash(coreDataModels: target.coreDataModels)
+        let targetActionsHash = try targetActionsContentHasher.hash(targetActions: target.actions)
+        let stringsToHash = [sourcesHash,
+                             target.name,
+                             target.platform.rawValue,
+                             target.product.rawValue,
+                             target.bundleId,
+                             target.productName,
+                             resourcesHash,
+                             coreDataModelHash,
+                             targetActionsHash]
+        //TODO: hash headers, platforms, version, entitlements, info.plist, target.environment, target.filesGroup, targetNode.settings, targetNode.project, targetNode.dependencies ,targetNode.target.dependencies
 
-    fileprivate func hash(string: String) throws -> String {
-        guard let hash = string.checksum(algorithm: .md5) else {
-            throw ContentHashingError.stringHashingFailed(string)
-        }
-        return hash
+        return try contentHasher.hash(stringsToHash)
     }
 
-    fileprivate func hash(strings: [String]) throws -> String {
-        guard let joinedHash = strings.joined().checksum(algorithm: .md5) else {
-            throw ContentHashingError.stringHashingFailed(strings.joined())
+    private func hash(headers: Headers) throws -> String {
+        let hashes = try (headers.private + headers.project + headers.project).map { path in
+            try contentHasher.hash(fileAtPath: path)
         }
-        return joinedHash
+        return try contentHasher.hash(hashes)
     }
 
-    fileprivate func md5(of source: Target.SourceFile) throws -> String {
-        guard let sourceData = try? fileHandler.readFile(source.path) else {
-            throw ContentHashingError.fileNotFound(source.path)
-        }
-        guard let hash = sourceData.checksum(algorithm: .md5) else {
-            throw ContentHashingError.fileHashingFailed(source.path)
-        }
-        return hash
+    private func hash(resources: [FileElement]) throws -> String {
+        let paths = resources.map { $0.path }
+        let hashes = try paths.map { try contentHasher.hash(fileAtPath: $0) }
```

Comment:
You are iterating paths twice, once to get the `.path` argument, and the other one to get the hash. I'd recommend doing only one iteration:

```swift
let hashes = try paths.map { try contentHasher.hash(fileAtPath: $0.path) }
```
Not a big deal if the list of resources is small, but it can be a performance bottleneck if the list is considerably large.

Reply:
very true, thanks! updated